iModelCore/ECDb/ECDb/RelationshipClassMap.cpp

```diff
@@ -1001,6 +1001,36 @@ ClassMappingStatus RelationshipClassLinkTableMap::_Map(ClassMappingContext& ctx)
     }
 
 //---------------------------------------------------------------------------------------
+//@bsimethod
+//---------------------------------------------------------------------------------------
+ClassMappingStatus RelationshipClassLinkTableMap::_UpdateDefaultIndexes(ClassMappingContext& ctx)
+    {
+    // Replace the index created in BisCore:ElementRefersToElements
+    if (Utf8String(ctx.GetClassMappingInfo().GetClass().GetFullName()).Equals("BisCore:ElementRefersToElements") && GetPrimaryTable().FindColumn("MemberPriority") != nullptr)
+        {
+        if (SUCCESS != ctx.GetImportCtx().GetSchemaManager().GetDbSchema().LoadIndexDefs())
+            return ClassMappingStatus::Error;
+
+        // Drop the old index and remove the entry from the table
+        if (GetPrimaryTable().RemoveIndexDef("uix_bis_ElementRefersToElements_sourcetargetclassid"))
+            {
+            if (!ctx.GetImportCtx().IsSynchronizeSchemas())
+                {
+                if (BE_SQLITE_OK != ctx.GetImportCtx().GetECDb().ExecuteSql(SqlPrintfString("DELETE FROM main." TABLE_Index " WHERE Name = 'uix_bis_ElementRefersToElements_sourcetargetclassid'")))
+                    return ClassMappingStatus::Error;
+                }
+
+            if (BE_SQLITE_OK != ctx.GetImportCtx().GetECDb().ExecuteDdl("DROP INDEX IF EXISTS [uix_bis_ElementRefersToElements_sourcetargetclassid]"))
+                return ClassMappingStatus::Error;
+            }
+
+        // Add the new index introduced in Profile version 4.0.0.5 if not already present
+        if (!GetPrimaryTable().ContainsIndex("uix_bis_ElementRefersToElements_sourcetargetclassid_memberpriority"))
+            AddIndex(ctx.GetImportCtx(), RelationshipIndexSpec::SourceAndTargetAndClassId, true);
+        }
+    return ClassMappingStatus::Success;

```

Comment:
This logic cannot exist in ECDb.  ECDb must look at the mapping CAs applied to BisCore.ElementRefersToElements to determine that it should change the index

Reply:
I second that, having a conditional code on the content of incoming data is a red flag for the maintainer. You need to remove this.